packages/common/src/api/tan-query/useUnfollowUser.ts

```diff
@@ -0,0 +1,201 @@
+import { Id } from '@audius/sdk'
+import { useMutation, useQueryClient } from '@tanstack/react-query'
+import { useDispatch, useSelector } from 'react-redux'
+
+import { useAudiusQueryContext } from '~/audius-query/AudiusQueryContext'
+import { useAppContext } from '~/context/appContext'
+import { Kind } from '~/models'
+import { Name, FollowSource } from '~/models/Analytics'
+import { Feature } from '~/models/ErrorReporting'
+import { ID } from '~/models/Identifiers'
+import { Track, isContentFollowGated } from '~/models/Track'
+import { UserMetadata } from '~/models/User'
+import { update } from '~/store/cache/actions'
+import { getTracks } from '~/store/cache/tracks/selectors'
+import { CommonState } from '~/store/commonStore'
+import { removeFolloweeId, revokeAccess } from '~/store/gated-content/slice'
+
+import { QUERY_KEYS } from './queryKeys'
+import { useCurrentUserId } from './useCurrentUserId'
+import { getUserQueryKey } from './useUser'
+import { getUserByHandleQueryKey } from './useUserByHandle'
+
+type UnfollowUserParams = {
+  followeeUserId: ID | null | undefined
+  source: FollowSource
+}
+
+type MutationContext = {
+  previousUser: UserMetadata | undefined
+  previousAccountUser: UserMetadata | undefined
+}
+
+export const useUnfollowUser = () => {
+  const { audiusSdk, reportToSentry } = useAudiusQueryContext()
+  const queryClient = useQueryClient()
+  const dispatch = useDispatch()
+  const {
+    analytics: { track, make }
+  } = useAppContext()
+  const { data: currentUserId } = useCurrentUserId()
+  const tracks = useSelector((state: CommonState) => getTracks(state, {}))
+
+  return useMutation({
+    mutationFn: async ({ followeeUserId, source }: UnfollowUserParams) => {
+      if (!currentUserId || !followeeUserId) {
+        return
+      }
+
+      const sdk = await audiusSdk()
+      await sdk.users.unfollowUser({
+        userId: Id.parse(currentUserId),
+        followeeUserId: Id.parse(followeeUserId)
+      })
+
+      // Handle gated content
+      dispatch(removeFolloweeId({ id: followeeUserId }))
+
+      // Revoke access to follow-gated tracks
+      if (tracks) {
```

Comment:
intense, do we need to do the affirmative during the useFollowUser?

Reply:
should we do all this en the "onMutate/onError" function for optimistic + rollback?

Reply:
I don't think so. This removes signatures we've acquired earlier in the session so access doesn't persist in the track metadata so for tracks where we don't have access yet, we'd try to sign before streaming? I don't think this actually affects the UI, it just purges stale signatures.

But honestly, I'm not sure we even need to do any of this. Is it a problem that the user retains access for the remainder of the session? Especially for a gate like follows which is pretty easy to acquire

Reply:
let's huddle on this today? imo it might be nice to at minimum just have the saga take care of this for now, and when we want implement signature stuff through tan query we can approach this? how you have it now is fine too, just want to get your thoughts

Reply:
yeah I'm moving it back to sagas rn